src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                        continue;
+                    }
+                    tempReferences.Add(MetadataReference.CreateFromFile(resolvedAssemblyPath));
+                }
+                return tempReferences;
+            }
+            else
+            {
+                return s_defaultAssemblies.Value;
+            }
         }
 
         private void WriteTypes(Assembly assembly)
         {
             WriteObject(assembly.GetTypes(), true);
         }
 
-        private void CompileSourceToAssembly(string source)
+        #endregion LoadAssembly
+
+        #region SourceCodeProcessing
+
+        private OutputKind OutputAssemblyTypeToOutputKind(OutputAssemblyType outputType)
         {
-            CSharpParseOptions parseOptions;
-            if (Language == Language.CSharp)
+            switch (outputType)
             {
-                parseOptions = new CSharpParseOptions();
+                case OutputAssemblyType.Library:
+                    return OutputKind.DynamicallyLinkedLibrary;
+                case OutputAssemblyType.ConsoleApplication:
+                    return OutputKind.ConsoleApplication;
+                case OutputAssemblyType.WindowsApplication:
+                    return OutputKind.WindowsApplication;
+                default:
+                    throw new ArgumentOutOfRangeException("outputType");
+            }
+        }
+
+        private static CSharpCommandLineArguments CSharpDefaultParse(IEnumerable<string> args, string baseDirectory = "./", string sdkDirectory = null, string additionalReferenceDirectories = null)
+        {
+            sdkDirectory = sdkDirectory ?? s_defaultSdkDirectory;
+            return CSharpCommandLineParser.Default.Parse(args, baseDirectory, sdkDirectory, additionalReferenceDirectories);
+        }
+
+        private static VisualBasicCommandLineArguments VisualBasicDefaultParse(IEnumerable<string> args, string baseDirectory = "./", string sdkDirectory = null, string additionalReferenceDirectories = null)
+        {
+            sdkDirectory = sdkDirectory ?? s_defaultSdkDirectory;
+            return VisualBasicCommandLineParser.Default.Parse(args, baseDirectory, sdkDirectory, additionalReferenceDirectories);
+        }
+
+        private void CSharpSourceCodeProcessing()
+        {
+            if (ExtendedOptions != null)
+            {
+                var arguments = CSharpDefaultParse(args: ExtendedOptions);
+
+                HandleCompilerErrors(arguments.Errors);
+
+                _csharpParseOptions = arguments.ParseOptions;
+                _csharpCompilationOptions = arguments.CompilationOptions.WithOutputKind(OutputAssemblyTypeToOutputKind(OutputType));
+
+                if (WarnAsError)
+                {
+                    _csharpCompilationOptions = arguments.CompilationOptions.WithGeneralDiagnosticOption(s_generalDiagnosticOption);
+                }
+
+                _emitOptions = arguments.EmitOptions;
```

Comment:
Same here, `_emitOptions` can be made a local variable, by adding another parameter to `CSharpCompileToAssembly()` as `EmitOption emitOption` and also pass the emit option to `DoEmitAndLoadAssemply(Compilation)`.

Reply:
Fixed